### PR TITLE
LDAP Group Provider Support

### PR DESCRIPTION
**Pull request by coryjmaklin**
This PR adds the code necessary to reference LDAP groups in Trino ACLs.
### Context

We're using file based ACLs to authorize users connected to Trino.
https://trino.io/docs/current/security/file-system-access-control.html

In the documentation, it's mentioned "For group-based rules to match, users need to be assigned to groups by a Group provider."

In other words, we have to create a group provider to specify the following:

```
{
"group": "trino_admins",
"privileges": ["SELECT", "INSERT", "DELETE", "OWNERSHIP"]
}
```

There exists a trino-password-authenticators plugin that takes care of authenticating users using their LDAP credentials.
https://github.com/trinodb/trino/tree/master/plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap

In the same plugin, there's already have a file based Group Provider.
https://github.com/trinodb/trino/blob/master/plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/PasswordAuthenticatorPlugin.java

There are open issues asking to add this as a feature.
https://github.com/trinodb/trino/issues/6824
https://github.com/trinodb/trino/issues/2919

### Implementation
- We move the common functions into a utility class.
- We rename the LdapAuthenticatorClient to LdapClient because Authenticator is a type of Trino plugin, and we're re-purposing the class for the Group Provider plugin.
- We know that a LDAP group will have a common name (i.e. `cn`) because it's required by the definition
- When retrieving the "memberof" attribute, the 'O' must be lowercase. When specifying that we want it in the search results, we the 'O' must be uppercase.

### Testing

- Configure the memberOf overlay in your LDAP instance

- Create a group-provider.properties with the following properties:

```
group-provider.name=ldap
ldap.allow-insecure=
ldap.url=
ldap.user-base-dn=
ldap.user-bind-pattern=
ldap.bind-dn=
ldap.bind-password=
```
- Connect to Trino and verify the ACLs are applied accordingly.

### CODE REVIEW

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    private final Optional<String> userBaseDistinguishedName;
+    private final Optional<String> bindDistinguishedName;
+    private final Optional<String> bindPassword;
+
+    @Inject
+    public LdapGroupProvider(LdapClient client, LdapConfig ldapConfig)
+    {
+        this.client = requireNonNull(client, "client is null");
+
+        this.userBindSearchPatterns = ldapConfig.getUserBindSearchPatterns();
```
Comment (cccs-tom) [inline plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:47]: Should we perform a requireNonNull() check on ldapConfig before reaching this point? Or are we just letting the NPE happen in that case?

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/JdkLdapClient.java [hunk 2]
```diff
@@ -106,11 +111,45 @@ public boolean isGroupMember(String searchBase, String groupSearch, String conte
...
+                        for (Rdn rdn : ln.getRdns()) {
+                            if (rdn.getType().equalsIgnoreCase("cn")) {
+                                groupNames.add(rdn.getValue().toString());
+                                break;
+                            }
+                        }
+                    }
+                }
+            }
+            search.close();
```
Comment (cccs-tom) [inline plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/JdkLdapClient.java:139]: Nitpick: This probably doesn't need to be closed explicitly, since it's managed by the try-with-resources.

**Review by cccs-tom — COMMENTED** [review]
I flagged a couple of minor things, but overall, lgtm!

**Comment by cccs-mjs** [comment]
Typo in PR description: `necessarily` -> `necessary`

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/JdkLdapClient.java [hunk 3]
```diff
@@ -49,16 +54,16 @@
 import static javax.naming.Context.SECURITY_CREDENTIALS;
 import static javax.naming.Context.SECURITY_PRINCIPAL;
 
-public class JdkLdapAuthenticatorClient
-        implements LdapAuthenticatorClient
+public class JdkLdapClient
```
Comment (cccs-mjs) [inline plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/JdkLdapClient.java:57]: What's the reason for renaming the class?  Could add to PR description.

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapClient.java [hunk 4]
```diff
@@ -17,7 +17,7 @@
 
 import java.util.Set;
 
-public interface LdapAuthenticatorClient
+public interface LdapClient
```
Comment (cccs-mjs) [inline plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapClient.java:20]: What's the reason for renaming the interface?  Could add to PR description.

**Review by cccs-mjs — COMMENTED** [review]
Seems good.  In the PR description it'd be useful to provide a sentence or two about the proposed code changes in general terms (e.g. new classes, name changes, moving shared methods out `LdapUtil` etc)

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/JdkLdapClient.java (hunk 2 above)
Comment (coryjmaklin) [reply plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/JdkLdapClient.java]: Updated

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java (hunk 1 above)
Comment (coryjmaklin) [reply plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:47]: I believe it's unnecessary because unlike the client, we're not keeping a reference, and we check the values afterwards.

**Comment by coryjmaklin** [comment]
> 
> 
> Typo in PR description: `necessarily` -> `necessary`

Updated

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapClient.java (hunk 4 above)
Comment (coryjmaklin) [reply plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapClient.java:20]: I added an explanation in the PR description. Essentially, Authenticator is a type of Trino plugin, but we're re-purposing the client for Group Provider plugin as well.

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/JdkLdapClient.java [hunk 5]
```diff
@@ -106,11 +111,45 @@ public boolean isGroupMember(String searchBase, String groupSearch, String conte
...
 
+    @Override
+    public Set<String> lookupUserGroups(String searchBase, String searchFilter, String contextUserDistinguishedName, String contextPassword)
+            throws NamingException
+    {
+        try (CloseableContext context = createUserDirContext(contextUserDistinguishedName, contextPassword); CloseableSearchResults search = searchContext(searchBase, searchFilter, context)) {
+            ImmutableSet.Builder<String> groupNames = ImmutableSet.builder();
+            if (search.hasMore()) {
+                Attributes attributes = search.next().getAttributes();
+                Attribute memberOfAttribute = attributes.get("memberof");
```
[the remark this answers is not in the document]
Comment (coryjmaklin) [reply plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/JdkLdapClient.java:122]: For whatever reason, `memberof` has to have a lowercase 'O' on line 122, and uppercase on line 151. If you don't it returns null.

**Comment by rimolive** [comment]
Any chance to help in this PR to have it merged asap?

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+            String searchBase = userBaseDistinguishedName.orElseThrow();
+            try {
+                return client.lookupUserGroups(searchBase, userDistinguishedName, bindDistinguishedName.orElseThrow(), bindPassword.orElseThrow());
+            }
+            catch (NamingException e) {
+                log.debug(e, "Authentication failed for user [%s], %s", user, e.getMessage());
+                throw new RuntimeException("Authentication error");
+            }
+        }
+        return null;
```
Comment (electrum) [inline plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:80]: This should return an empty set
```suggestion
        return ImmutableSet.of();
```

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapAuthenticator.java [hunk 7]
```diff
@@ -144,7 +141,7 @@ private Principal authenticateWithUserBind(Credential credential)
     private Principal authenticateWithBindDistinguishedName(Credential credential)
     {
         String user = credential.getUser();
-        if (containsSpecialCharacters(user)) {
+        if (LdapUtil.containsSpecialCharacters(user)) {
```
Comment (kokosing) [inline plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapAuthenticator.java:144]: extraction of `LdapUtil` can go as separate commit too.

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+                "Both bind distinguished name and bind password must be provided");
+        checkArgument(
+                !userBindSearchPatterns.isEmpty(),
+                "User bind search pattern must be provided");
+    }
+
+    @Override
+    public Set<String> getGroups(String user)
+    {
+        if (LdapUtil.containsSpecialCharacters(user)) {
```
Comment (kokosing) [inline plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:66]: static import

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+            throw new AccessDeniedException("Username contains a special LDAP character");
+        }
+        for (String userBindSearchPattern : userBindSearchPatterns) {
+            String userDistinguishedName = LdapUtil.replaceUser(userBindSearchPattern, user);
+            String searchBase = userBaseDistinguishedName.orElseThrow();
+            try {
+                return client.lookupUserGroups(searchBase, userDistinguishedName, bindDistinguishedName.orElseThrow(), bindPassword.orElseThrow());
```
Comment (kokosing) [inline plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:73]: We are breaking a loop in first iteration and so we are ignoring rest of `userBindSearchPatterns`. Is this intentional? If so, why? Can you please add a comemnt?

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProviderFactory.java [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    @Override
+    public GroupProvider create(Map<String, String> config)
+    {
+        Bootstrap app = new Bootstrap(
+                binder -> {
+                    configBinder(binder).bindConfig(LdapConfig.class);
+                    binder.bind(LdapGroupProvider.class).in(Scopes.SINGLETON);
+                    binder.bind(LdapClient.class).to(JdkLdapClient.class).in(Scopes.SINGLETON);
```
[the remark this answers is not in the document]
Comment (kokosing) [reply plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProviderFactory.java:42]: It is in separate guice context, so I think it is not an issue.

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapUtil.java [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.password.ldap;
+
+import com.google.common.base.CharMatcher;
+
+public class LdapUtil
```
Comment (kokosing) [inline plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapUtil.java:18]: final

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapUtil.java [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+{
+    private static final CharMatcher SPECIAL_CHARACTERS = CharMatcher.anyOf(",=+<>#;*()\"\\\u0000");
+    private static final CharMatcher WHITESPACE = CharMatcher.anyOf(" \r");
+
+    private LdapUtil()
```
Comment (kokosing) [inline plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapUtil.java:23]: `private LdapUtil() {}`

File: plugin/trino-password-authenticators/src/test/java/io/trino/plugin/password/ldap/TestLdapClient.java [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+import io.trino.plugin.password.Credential;
+
+import javax.naming.NamingException;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+
+public class TestLdapClient
```
[the remark this answers is not in the document]
Comment (kokosing) [reply plugin/trino-password-authenticators/src/test/java/io/trino/plugin/password/ldap/TestLdapClient.java:27]: Notice that we have `SinglenodeLdap` or `SinglenodeLdapBindDn` environments already created. 

Basically to create product tests, you need to add a test `io.trino.tests.product.jdbc.TestLdapTrinoJdbc` and add to the product test environment configuration for LDAP group provider.

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java (hunk 9 above)
Comment (hashhar) [reply plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java:73]: Reminder to keep https://github.com/trinodb/trino/pull/8134 in mind. All patterns must be tried and only the last AccessDeniedException must be bubbled up.

**Comment by MiguelWeezardo** [comment]
@coryjmaklin how can we help to get this merged?

**Comment by kokosing** [comment]
@coryjmaklin Would you mind if we would take over this pull request?

**Comment by coryjmaklin** [comment]
> 
> 
> @coryjmaklin Would you mind if we would take over this pull request?

Sorry. I don't have time to work on it at present. If you want to take it over, by all means.

**Comment by rimolive** [comment]
Anyone knows why am I getting this error in trino-main tests?

`[ERROR] Tests run: 7872, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 969.778 s <<< FAILURE! - in TestSuite
[ERROR] io.trino.util.TestTimeZoneUtils.testNamedZones  Time elapsed: 0.037 s  <<< FAILURE!
java.lang.IllegalArgumentException: The datetime zone id 'Pacific/Kanton' is not recognised
        at org.joda.time.DateTimeZone.forID(DateTimeZone.java:247)
        at io.trino.util.TestTimeZoneUtils.testNamedZones(TestTimeZoneUtils.java:51)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
        at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
        at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
        at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
        at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
        at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)`

I'm using OpenJDK 11.0.13 to run maven build

**Comment by MiguelWeezardo** [comment]
> Anyone knows why am I getting this error in trino-main tests?
> 
> `[ERROR] Tests run: 7872, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 969.778 s <<< FAILURE! - in TestSuite [ERROR] io.trino.util.TestTimeZoneUtils.testNamedZones Time elapsed: 0.037 s <<< FAILURE! java.lang.IllegalArgumentException: The datetime zone id 'Pacific/Kanton' is not recognised at org.joda.time.DateTimeZone.forID(DateTimeZone.java:247) at io.trino.util.TestTimeZoneUtils.testNamedZones(TestTimeZoneUtils.java:51) at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) at java.base/java.lang.reflect.Method.invoke(Method.java:566) at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104) at org.testng.internal.Invoker.invokeMethod(Invoker.java:645) at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851) at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177) at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129) at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112) at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) at java.base/java.lang.Thread.run(Thread.java:829)`
> 
> I'm using OpenJDK 11.0.13 to run maven build

Looks like your JDK knows of a zone which isn't yet supported by joda-time version Trino uses.
I see a couple of possible solutions:
* upgrade joda-time
* downgrade to OpenJDK 11.0.11
* add an exception to `TestTimeZoneUtils.testNamedZones`

**Comment by rimolive** [comment]
@kokosing @MiguelWeezardo Please check updated PR: #10116

**Comment by kokosing** [comment]
Closing in favor #10116. Thank @rimolive
